### DISCUSSION TITLE
Nested transactions

### DISCUSSION THREAD

**Discussion by powellnorma** [piccolo-orm/piccolo#729] [General]
Does it make sense to just "ignore" nested transactions instead of throwing an `TransactionError`? Or could this by dangerous?

https://github.com/piccolo-orm/piccolo/blob/d5123e94337fe4ee05816b890cfdbcc4a897009c/piccolo/engine/postgres.py#L142-L146

**Comment by dantownsend** [comment]
Yeah, it's a good point.

Some ORMs just ignore it if you try creating a new transaction inside an existing transaction, unless you explicitly ask for a sub transaction (e.g. [Ruby on Rails](https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#module-ActiveRecord::Transactions::ClassMethods-label-Nested+transactions)).

Others will automatically create a sub transaction using `SAVEPOINT` (e.g. Django).

We could probably add sub transaction support to Piccolo fairly easily. I didn't add it originally probably due to time constraints.

Which behaviour / design do you think is preferable?

**Reply by dantownsend** [reply]
It could be configurable I guess.

```python
class NestedTransactionBehaviour(Enum):
    use_savepoint = 'use_savepoint'
    ignore = 'ignore'
    raise_exception = 'raise_exception'

async with b.transaction(nested=NestedTransactionBehaviour.savepoint):
    ...
```

**Reply by powellnorma** [reply]
Just saw your reply now - I think just ignoring it would be fine. But I should probably say that I haven't heard of "sub transactions" until now. I also am not sure what they do. Is it possible that a inner transaction commits but the outer does not? Because otherwise, I don't see how a sub transaction changes the behavior - I would have to do some research

**Reply by dantownsend** [reply]
With sub-transactions it uses SQL's `SAVEPOINT`. It allows you to roll back part of the transaction. It's not something I use much, but is interesting.

**Reply by powellnorma** [reply]
How would/can one make use of this in piccolo? Can one manually roll back a transaction? Until now I more thought of transactions as a "exception handler" that rolls back everything, and when an inner transaction would throw, that would also pass the "exception" to the parent transaction. But it sounds like it is possible to make the inner transaction "swallow" exceptions(?)

**Reply by dantownsend** [reply]
Throwing an exception within a transaction will stop the transaction.

```python
async with DB.transaction():
    await Band.insert(Band(name='Test'))
    raise Exception('Roll back')
```

The exception is still propagated though, so if there were nested transactions we would have to catch the exception from the sub transaction to prevent it from rolling the parent transaction back as well.

```python
async with DB.transaction():
    await Band.insert(Band(name='Test'))
    
    try:
        async with DB.transaction():
            await Band.insert(Band(name='Test 2'))
            raise Exception('Roll back')
    except Exception:
        pass
```

One thing we could do, which would be fairly simple to implement, is have the async context manager return the transaction, so we can roll back without having to throw an exception.

```python
async with DB.transaction():
    await Band.insert(Band(name='Test'))
    
    async with DB.transaction() as sub_transaction:
        await Band.insert(Band(name='Test 2'))
        await sub_transaction.rollback()
```

**Comment by dantownsend** [comment]
There's a PR which implements savepoints based off the conversation here.

https://github.com/piccolo-orm/piccolo/pull/778

**Comment by dantownsend** [comment]
> Does it make sense to just "ignore" nested transactions instead of throwing an TransactionError? Or could this by dangerous?

I'll change the behaviour to do this in the next Piccolo release - you're right it makes more sense. There will be an option to use the old behaviour, but I think the default should be as you suggest.
